### ISSUE TITLE
Hotkeys and toolbar plugins defined by in-game scripts are not saved with the game

### ISSUE THREAD

**Issue by bunnybot** [widelands/widelands#7031]
[<img align='left' src='https://codeberg.org/avatars/593dc43e2444c25db9481960e5b693ab' alt='tothxa' height='48' />](https://codeberg.org/tothxa)*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/5390)*
*Created on Sun May 17 00:07:46 CEST 2026 by [**Tóth András (tothxa)**](https://codeberg.org/tothxa)*

---

### Describe the Bug

Any kind of add-ons are allowed to define keyboard shortcuts, and they can attach lua scripts to them in their code. However, the shortcuts are tied to the `MapView`, not the game, and they are not saved when the game is saved, so the defined hotkeys become unavailable after reloading the game if the script is not re-ran, e.g. win condition scripts. Toolbar plugins behave the same.

There is no easy workaround either, because there is no way to query whether a keyboard shortcut has a script attached, nor what toolbar plugins are active. There's also no way to automatically run a script on game loading.

### To Reproduce

1. Install the [attached add-on](https://codeberg.org/attachments/766562cb-3943-4f44-9c14-181e9bf12bb3)
2. Start a game with the add-on's "UI demo" win condition
3. Check that the hotkey and the toolbar plugin work
4. Save and reload the game
5. See error: the hotkey doesn't work any more, and the toolbar plugin is missing

### Expected Behavior

There should be a way to register keyboard shortcuts and toolbar plugins as persistent, so they are saved with the game and automatically re-registered on game loading.

In any way, it should also be possible to access active keyboard shortcuts and toolbar plugins from lua, similar to plugin timers. (this would enable a workaround to the above too by polling the defined items regularly, and re-register them when they go missing, but it's a sub-optimal solution to the problem)

A possible alternative would be to allow defining a script to be run on game loading, but I'm afraid that would open up undesired possibilities.

### Operating System

any

### Widelands Version

1.4~git27263 (56a55ad@master)

**Comment by bunnybot** [comment]
[<img align='left' src='https://codeberg.org/avatars/a4ad2f9a2504f052397d8ba82ed48ba1' alt='Nordfriese' height='48' />](https://codeberg.org/Nordfriese)*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/5390#issuecomment-15119319)*
*On Sun May 17 11:08:17 CEST 2026, [**Benedikt Straub (Nordfriese)**](https://codeberg.org/Nordfriese) wrote:*

---

Hm, this is working as designed from my point of view, shortcuts and toolbar plugins are intended to be used only with add-ons of category UI Plugin.

However +1 to a more extensive Lua API for plugin elements.
And I would also say +1 for a way for add-ons to know when the game was reloaded, this _is_ a sore point for scripts that want to interface with both user interface and game state. Running a script with a fixed name if it exists directly after loading is complete should be safe AFAICS so long as the script behaves the same way all add-on scripts should; which undesired possibilities do you see here?

**Comment by bunnybot** [comment]
[<img align='left' src='https://codeberg.org/avatars/593dc43e2444c25db9481960e5b693ab' alt='tothxa' height='48' />](https://codeberg.org/tothxa)*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/5390#issuecomment-15150921)*
*On Sun May 17 15:36:01 CEST 2026, [**Tóth András (tothxa)**](https://codeberg.org/tothxa) wrote:*

---

<@>Nordfriese wrote in https://codeberg.org/wl/widelands/issues/5390#issuecomment-15119319:

> Hm, this is working as designed from my point of view, shortcuts and toolbar plugins are intended to be used only with add-ons of category UI Plugin.

Well, they are allowed for all add-on types, and there are possible uses too. In my case, I wrote a win condition with a composite score, so I created a window with detailed score info that can be opened by a hotkey or from the menu. But the status reports in official win conditions could be made available in a similar way too.

I can also imagine scenarios that could make use of these, and maybe even starting conditions that enable some special functionality that needs a UI.

None of these survives a save-reload cycle, and I think that's a quite serious bug for these use cases.


> I would also say +1 for a way for add-ons to know when the game was reloaded, this _is_ a sore point for scripts that want to interface with both user interface and game state. Running a script with a fixed name if it exists directly after loading is complete should be safe AFAICS so long as the script behaves the same way all add-on scripts should; which undesired possibilities do you see here?

I wouldn't want to risk changes in the game logic flow depending on whether it's continuous play or saved and reloaded.

**Comment by bunnybot** [comment]
[<img align='left' src='https://codeberg.org/avatars/593dc43e2444c25db9481960e5b693ab' alt='tothxa' height='48' />](https://codeberg.org/tothxa)*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/5390#issuecomment-15173091)*
*On Mon May 18 00:04:23 CEST 2026, [**Tóth András (tothxa)**](https://codeberg.org/tothxa) wrote:*

---

The plot thickens:

I tried to create a separate UI add-on to register the hotkey and the toolbar plugin, but it's really hard to properly test for the win condition, because the `init.lua` of UI add-ons is run before the game is set up.

The least horrible solution would be to run a plugin timer, because its actual first run is always after the game has been set up properly, but that's not an option for v1.3.1 where all plugin timers were infinite, because then we're back to square one: we can't tell whether we've already registered the hotkey and the toolbar plugin.

TLDR: Couldn't UI add-on scripts be run *after* the game is properly loaded/started?

**Comment by bunnybot** [comment]
[<img align='left' src='https://codeberg.org/avatars/593dc43e2444c25db9481960e5b693ab' alt='tothxa' height='48' />](https://codeberg.org/tothxa)*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/5390#issuecomment-15173487)*
*On Mon May 18 00:14:56 CEST 2026, [**Tóth András (tothxa)**](https://codeberg.org/tothxa) wrote:*

---

Oh, yes, and some more bugs I'm too exhausted to report separately:

 - during game setup (when UI add-on scripts are run) the win condition is reported as the localised form of "Not set"
 - trying to use `sleep()` in UI add-on scripts causes a segfault in loaded games, I guess because it sets up a game logic command that's due in the past. Running the scripts after game setup would probably solve this too.

**Comment by bunnybot** [comment]
[<img align='left' src='https://codeberg.org/avatars/593dc43e2444c25db9481960e5b693ab' alt='tothxa' height='48' />](https://codeberg.org/tothxa)*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/5390#issuecomment-15964319)*
*On Wed May 27 18:06:30 CEST 2026, [**Tóth András (tothxa)**](https://codeberg.org/tothxa) wrote:*

---

<@>tothxa wrote in https://codeberg.org/wl/widelands/issues/5390#issuecomment-15173091:

> The least horrible solution would be to run a plugin timer, because its actual first run is always after the game has been set up properly, but that's not an option for v1.3.1 where all plugin timers were infinite, because then we're back to square one: we can't tell whether we've already registered the hotkey and the toolbar plugin.

See horrible work-around in https://codeberg.org/wl/wl_addons_server/src/addons/win_cond_mil_inf_ui.wad/init.lua
Turns out any status variables from UI add-on scripts are saved with the game state, even local ones. So that goes the other way, though I can see that it's nearly impossible to fix. (maybe if we ran UI scripts in separate lua contexts, but then they wouldn't be able to see the game state)
